client/go/outline/vpn/nmconn_linux.go

```diff
@@ -34,73 +41,110 @@ type nmConnectionOptions struct {
-
-	if err = dev.SetPropertyManaged(true); err != nil {
-		return nil, errSetupVPN("failed to manage tun device", err, "dev", dev.GetPath(), "api", "NetworkManager")
-	}
-	slog.Debug("NetworkManager now manages the tun device", "dev", dev.GetPath())
-
 	props := make(map[string]map[string]interface{})
 	configureCommonProps(props, opts)
 	configureTUNProps(props)
 	configureIPv4Props(props, opts)
 	slog.Debug("populated NetworkManager connection settings", "settings", props)
 
-	// The previous SetPropertyManaged call needs some time to take effect (typically within 50ms)
-	for retries := 20; retries > 0; retries-- {
+	dev, err := nm.GetDeviceByIpIface(opts.TUNName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to locate TUN device in NetworkManager: %w", err)
+	}
+
+	for retries := nmAPIRetryCount; retries > 0; retries-- {
 		slog.Debug("trying to create NetworkManager connection for tun device...", "dev", dev.GetPath())
-		ac, err = nm.AddAndActivateConnection(props, dev)
+		c.ac, err = nm.AddAndActivateConnection(props, dev)
 		if err == nil {
+			slog.Info("successfully created NetworkManager connection", "conn", c.ac.GetPath())
 			break
 		}
 		slog.Debug("failed to create NetworkManager connection, will retry later", "err", err)
-		time.Sleep(50 * time.Millisecond)
+		time.Sleep(nmAPIRetryDelay)
 	}
-	if err != nil {
-		return ac, errSetupVPN("failed to create connection", err, "dev", dev.GetPath(), "api", "NetworkManager")
+	return c, err
+}
+
+func (c *nmConnection) Close() error {
+	if c.ac != nil {
+		if err := c.nm.DeactivateConnection(c.ac); err != nil {
+			slog.Warn("failed to deactivate NetworkManager connection", "err", err, "conn", c.ac.GetPath())
+		}
+		slog.Debug("deactivated NetworkManager connection", "conn", c.ac.GetPath())
 	}
-	return
+	return clearNMConnections(c.nm, c.name)
 }
 
-func closeNMConnection(nm gonm.NetworkManager, ac gonm.ActiveConnection) error {
+// clearNMConnections removes all NetworkManager connections with a given name.
+func clearNMConnections(nm gonm.NetworkManager, name string) error {
 	if nm == nil {
 		panic("a NetworkManager must be provided")
 	}
-	if ac == nil {
+	if name == "" {
 		return nil
 	}
 
-	if err := nm.DeactivateConnection(ac); err != nil {
-		slog.Warn("failed to deactivate NetworkManager connection", "err", err, "conn", ac.GetPath())
-	}
-	slog.Debug("deactivated NetworkManager connection", "conn", ac.GetPath())
-
-	conn, err := ac.GetPropertyConnection()
-	if err == nil {
-		err = conn.Delete()
-	}
+	slog.Debug("removing all NetworkManager connections with name ...", "name", name)
+	nmSettings, err := gonm.NewSettings()
 	if err != nil {
-		return errCloseVPN("failed to delete NetworkManager connection", err, "conn", ac.GetPath())
+		return fmt.Errorf("failed to connect to NetworkManager settings: %w", err)
 	}
-	slog.Info("NetworkManager connection deleted", "conn", ac.GetPath())
-
-	return nil
+	for retries := nmAPIRetryCount; retries > 0; retries-- {
```

Comment:
Can you move the loop into the `ListConnections` call instead, so the caller logic is clearer?

Same for `AddAndActivateConnection`

Reply:
Done.